### PR TITLE
Refactor plugin code

### PR DESCRIPTION
**Pull request by DIOHz0r**
Hi, @btry

This pull request is refered to issue #134, about refactoring the plugin code.

Best regards.

### CODE REVIEW

**Comment by ajsb85** [comment]
Hi, @DIOHz0r 

Too long commit message:

```
 refactor(invitation): method getFromDBbyName uses a second invalid pa…
…rameter
```

**Comment by DIOHz0r** [comment]
Hi @ajsb85,

The commit comment about invitation class was edited. I will be still working on this PR.

Regards.

File: inc/agent.class.php [hunk 1]
```diff
@@ -1577,6 +1584,7 @@ public function getOwner() {
    /**
     * Determine the enrollment method depending on input data
     * @param array $input
+    * @return int
```
Comment (btry) [inline inc/agent.class.php]: please use integer as you choos if elsewhere already

File: inc/fleet.class.php [hunk 2]
```diff
@@ -101,6 +102,8 @@ public function defineTabs($options = []) {
 
    /**
     * Show form for edition
+    * @param $ID
+    * @param array $options
     */
    public function showForm($ID, $options = []) {
```
Comment (btry) [inline inc/fleet.class.php]: type hint array for $options (becasue you did it earlier i nthis PR)

Comment (DIOHz0r) [reply inc/fleet.class.php]: understood

**Comment by DIOHz0r** [comment]
Hi,

Could @btry and @ajsb85 check the current progress so far?

Regards.

File: inc/agent.class.php (hunk 1 above)
Comment (ajsb85) [reply inc/agent.class.php]: @DIOHz0r Remember to participate in the discuss of the Pull Request 🤓

**Comment by ajsb85** [comment]
Hi, @DIOHz0r 
The ellipses in the commits are the signal that they are very long commits that you must correct

imagen

File: inc/task.class.php [hunk 3]
```diff
@@ -538,7 +533,8 @@ protected function buildMqttMessage($policiesToApply) {
          );
          if ($policyMessage === false) {
             // There is an error while applying the policy, continue with next one for minimal impact
-            // TODO: log something
+            Toolbox::logInFile('plugin_flyvemdm_inventory',
+               "Policy value '" . $policyToApply['value'] . "' not applied\n");
```
Comment (btry) [inline inc/task.class.php]: ok to implement logging, but maybe should we add more details : which policy ? which fleet ?

We don't have the fleet object here. Maybe introducing an exception would be better. (and add try / catch block at the level where the  fleet instance is available

**Comment by DIOHz0r** [comment]
Hi,

I'm goint to fix the comments that you made today. Thank you.

File: inc/policyfactory.class.php [hunk 4]
```diff
@@ -42,7 +42,9 @@ class PluginFlyvemdmPolicyFactory {
    /**
     * Create an empty policy by type
     *
-    * @param string $type
+    * @param PluginFlyvemdmPolicy $policyData
+    * @return null|PluginFlyvemdmPolicyInterface depending on the field type.
+    * @internal param string $type
```
[the remark this answers is not in the document]
Comment (DIOHz0r) [reply inc/policyfactory.class.php]: The IDE did it and I din't notice that.

**Comment by DIOHz0r** [comment]
Hi,

Could you review this advance again please?

File: inc/invitation.class.php [hunk 5]
```diff
@@ -293,10 +292,10 @@ public function getUser() {
    /**
     * get the enrollment URL of the agent
     * @param User $user Recipient of the QR code
-    * @param string $învitationToken Invitation token
+    * @param string $invitationToken Invitation token
```
Comment (btry) [inline inc/invitation.class.php]: good point ! By the way this means PHP handles accented chars in var names ??

File: inc/policyremoveapplication.class.php [hunk 6]
```diff
@@ -39,9 +39,10 @@
 class PluginFlyvemdmPolicyRemoveapplication extends PluginFlyvemdmPolicyBase implements PluginFlyvemdmPolicyInterface {
 
    /**
-    * @param string $properties
+    * @param PluginFlyvemdmPolicy $policy
     */
    public function __construct(PluginFlyvemdmPolicy $policy) {
+      parent::__construct($policy);
```
Comment (btry) [inline inc/policyremoveapplication.class.php]: you're right, this is safer

**Comment by btry** [comment]
Hi @DIOHz0r this PR loogs good to me so far.

Just one thing about the exception you introduced on my request. I need to check how GLPI can manage namespaces for exceptions, and create a class inherited from Excepion with a proper name. For now, let's leave it as is.

**Comment by DIOHz0r** [comment]
@btry OK, some more changes are comming and I think I'm done with this PR.

**Comment by DIOHz0r** [comment]
Hello,

Could you review this PR again?. It's finished.

Thank you.

**Comment by btry** [comment]
@DIOHz0r I added some commits to the branch to change the exception you introduced. Actually I tested the suggestion I talked earlier, and it worked, so I pushed. We need to limit namespaces to exceptions only; GLPI is not ready for namespaced itemtypes.

As I expect we will add more exceptions, I think this will help to not blow up the /inc folder.

I also harmonized the upgrade code to use a $table var everywhere. Before releasing 2.0.0 the 2 upgrade_*.php files shall merge.

Note that there is a CREATE TABLE statement and immediately followed by $migratoin->addKey() this is normal, and these addKey() shall be dropped only when we will prepare the release.

My review of your commits => OK, now please review mines, just in case.

File: install/upgrade/update_to_dev.php [hunk 7]
```diff
@@ -119,25 +122,29 @@ function plugin_flyvemdm_update_to_dev(Migration $migration) {
...
+   $table = 'glpi_plugin_flyvemdm_invitationlogs';
+   $migration->addKey($table, 'glpi_plugin_flyvemdm_invitationlogs', 'invitations_id');
 
    $table = 'glpi_plugin_flyvemdm_invitations';
    $migration->addKey($table, 'users_id', 'users_id');
    $migration->addKey($table, 'entities_id', 'entities_id');
    $migration->addKey($table, 'documents_id', 'documents_id');
 
-   $table = 'glpi_plugin_flyvemdm_taskstatuses';
-   $migration->addKey($table, 'plugin_flyvemdm_agents_id', 'plugin_flyvemdm_agents_id');
```
Comment (DIOHz0r) [inline install/upgrade/update_to_dev.php]: Are not necessary this keys?, the update_to_2_0_0.php file has them.

File: install/upgrade/update_to_dev.php [hunk 8]
```diff
@@ -119,25 +122,29 @@ function plugin_flyvemdm_update_to_dev(Migration $migration) {
...
-   $migration->addKey($table, 'plugin_flyvemdm_tasks_id', 'plugin_flyvemdm_tasks_id');
```
Comment (DIOHz0r) [inline install/upgrade/update_to_dev.php]: Are not necessary this keys?, the update_to_2_0_0.php file has them.

Comment (btry) [reply install/upgrade/update_to_dev.php]: actually I grouped the changes by table.
* $table affectation moved to L82
* addkey() calls moved to L98

Comment (DIOHz0r) [reply install/upgrade/update_to_dev.php]: ok then it looks good for me!.

**Comment by ajsb85** [comment]
Hi, @DIOHz0r 
You can add in the description that the PR close the issue #134
So, when the PR is merged, the issues is closed automatically 👍

**Pull request merged**